src/ImageRunModal.jsx

```diff
@@ -27,13 +28,107 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports.splice(idx, 1);
+            if (ports.length === 0)
+                ports.push({ containerPort: null, hostPort: null, protocol: 'TCP' });
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    addPort() {
+        this.setState(state => {
+            return { publish: [...state.publish, { containerPort: null, hostPort: null, protocol: 'TCP' }] };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    onPortChange(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports[idx][field] = value || null;
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
```

Comment:
Can you map a container's port X/TCP to one host port, and container port X/UDP to a *different* host port? Intuitively this shouldn't work. If not, then these functions should check this case. I take it you want to keep the array explicitly ordered, otherwise I'd suggest a containerPort → { hostPort, protocol } mapping as state. But this would lead to the UI jumping around, so let's keep the array.

Reply:
right now (almost) all the validation is actually done by podman itself. Do we want to start doing extensive validation ? The `.filter` based on ports is just to make sure we don't send any empty rows really, `null` there would be more explicit I guess.

Reply:
Also it's possible to have multiple host ports mapped to a single container port, so 127.1:8080->80/tcp and 127.1:8081->80/tcp for instance.

Reply:
Ah, very well. Does the dialog show a proper error on a port conflict? That's worth adding to the integration test, as that's the sort of thing that tends to silently break with changing versions.